### ISSUE TITLE
[Bug] Per-layer optimizers step on every backward, ignoring gradient accumulation

### ISSUE THREAD

**Issue by gss10282023** [jiaweizzhao/GaLore#78]
### Summary

With `galore_adamw8bit_per_layer`, the per-parameter hook calls `optimizer.step()`, `zero_grad()` and `scheduler.step()` on every microbatch, before the outer loop reaches its accumulation boundary:

```python
def optimizer_hook(p):
    ...
    optimizer_dict[p].step()
    optimizer_dict[p].zero_grad()
    scheduler_dict[p].step()
```

With `gradient_accumulation=4`, one logical update becomes four Adam updates on quarter-scaled losses. Dividing each loss by 4 does not make four Adam steps equal to one step on the summed gradient.

This is related to the step-counting discussion in #62. The single-GPU recipe in the README combines this optimizer with `--batch_size 16 --total_batch_size 512`, so the documented entry point runs with accumulation.

### Environment

Commit [2cc66f88](https://github.com/jiaweizzhao/GaLore/blob/2cc66f88cce189e505affbb91042a8e77f5bf4e9/torchrun_main.py), PyTorch 2.8.0+cu128, bitsandbytes 0.50.2, one RTX 5090.

### Steps to reproduce

Run `torchrun_main.py` with a two-layer LLaMA (hidden size 128, vocabulary 512), rank 32, four fixed sequences of length 32, `GaLoreAdamW8bit`, FP32, no gradient clipping. Do one logical update with `gradient_accumulation=1` and with `gradient_accumulation=4`, same data and initial state.

| After one logical update | GA 1 | GA 4 |
| --- | ---: | ---: |
| Optimizer steps per parameter | 1 | 4 |
| Scheduler steps per parameter | 1 | 4 |

All 21 parameters show this. The parameter-update relative L2 difference between the two runs is `0.9827118878`.

Deferring the same step/zero_grad/scheduler calls to the existing accumulation boundary brings the difference down to `0.0026727007`, which is within the noise of the 8-bit optimizer measured separately. The deferred version is bit-identical to the current code when `gradient_accumulation=1`. Reproduced on a second host.

### Expected behavior

`gradient_accumulation` should mean the same thing for the per-layer optimizer as for the others: one optimizer and scheduler step per logical update.

### Proposed fix

Have the hook only record that a parameter has a gradient, and run each pending parameter's optimizer step, zero_grad and scheduler step once at the existing update boundary. Gradients then stay in memory until that point.

Because of that, this change probably does not keep the memory footprint reported for the 7B setup; I have not measured it. If the per-layer path is meant to support only `gradient_accumulation=1`, rejecting larger values and updating the README recipe would also be a valid fix. The numbers above are from a small FP32 model with clipping disabled, chosen so the update order is the only difference.
